crates/astria-sequencer/src/app/mod.rs

```diff
@@ -781,16 +820,73 @@ impl App {
             .await
             .wrap_err("failed to get sudo address from state")?;
 
-        // convert tendermint id to astria address; this assumes they are
-        // the same address, as they are both ed25519 keys
-        let proposer_address = finalize_block.proposer_address;
+        let end_block = self.end_block(height.value(), sudo_address).await?;
+
+        // get deposits for this block from state's ephemeral cache and put them to storage.
+        let mut state_tx = StateDelta::new(self.state.clone());
+        let deposits_in_this_block = self.state.get_cached_block_deposits();
+        debug!(
+            deposits = %telemetry::display::json(&deposits_in_this_block),
+            "got block deposits from state"
+        );
+
+        state_tx
+            .put_deposits(&block_hash, deposits_in_this_block.clone())
+            .wrap_err("failed to put deposits to state")?;
+
+        // cometbft expects a result for every tx in the block, so we need to return a
+        // tx result for the commitments, even though they're not actually user txs.
+        //
+        // the tx_results passed to this function only contain results for every user
+        // transaction, not the commitment, so its length is len(txs) - 2.
+        let mut finalize_block_tx_results: Vec<ExecTxResult> = Vec::with_capacity(txs.len());
+        finalize_block_tx_results.extend(std::iter::repeat(ExecTxResult::default()).take(2));
+        finalize_block_tx_results.extend(tx_results);
 
-        let height = finalize_block.height;
-        let time = finalize_block.time;
-        let Hash::Sha256(block_hash) = finalize_block.hash else {
-            bail!("finalized block hash is empty; this should not occur")
+        let sequencer_block = SequencerBlock::try_from_block_info_and_data(
+            block_hash,
+            chain_id,
+            height,
+            time,
+            proposer_address,
+            txs,
+            deposits_in_this_block,
+        )
+        .wrap_err("failed to convert block info and data to SequencerBlock")?;
+        state_tx
+            .put_sequencer_block(sequencer_block)
+            .wrap_err("failed to write sequencer block to state")?;
+
+        // events that occur after end_block are ignored here;
+        // there should be none anyways.
+        let _ = self.apply(state_tx);
+
+        // use a dummy app hash here - the actual app hash will be filled out in finalize_block.
+        let dummy_app_hash = AppHash::default();
+
+        let finalize_block = abci::response::FinalizeBlock {
```

Comment:
It's surprising that this method changes the internal state of `self.finalize_block`, but the `finalize_block` takes a `finalize_block` argument and then mutates the `finalize_block` field that some other part of the stack changed.

Reply:
i could change this to be some intermediate type that doesn't have the app hash yet, which is turned into `finalize_block` response only in `finalize_block`. would that be easier to follow?

Reply:
I suppose it's fine. We should probably sit down and refactor of all of this. It looks like we have a bit of code duplication going on.